lib/build-components/build_components/__main__.py

```diff
@@ -0,0 +1,511 @@
+            name=dict(color="white", bold=True),
+            programname=dict(color="cyan"),
+            username=dict(color="yellow"),
+        ),
+    )
+
+    # This will be a VerboseLogger
+    logger = logging.getLogger(__name__)
+
+    # Provide the lamp.json file data to the build process
+    with open("lamp.json") as lamp_file:
+        lamp = json.load(lamp_file)
+
+    # Let's get to it!
+    build_components(
+        lamp=lamp,
+        tag_hash=os.getenv("HASH", "local"),
+        build_config=BuildConfig(),
+        components=sys.argv[1:],
+    )
+
+
+def build_components(
+    lamp: dict, tag_hash: str, build_config=BuildConfig, components: typing.List[str] = None
+):
+    """
+    Build each component for the project.
+
+    If components is empty, this will assume each directory in the components/ directory is a
+    component and will build each of them.
+
+    Each component will result in up to two images being built: the image to be used in whatever
+    intended use case the component fulfils, and a second image, identical to the first except
+    that its ENTRYPOINT and CMD instructions have been "cleared". This second image will only be
+    built if tag_hash is "local". The resulting images will be tagged as
+    ``{project}-{component}:{tag_hash}`` and ``{project}-{component}:editor``, respectively.
+
+    :param lamp: The data from the project's lamp.json file.
+    :param tag_hash: The build hash provided by ``aladdin build``.
+    :param build_config: The default values for general build settings.
+    :param components: The list of components to build, defaults to all of them.
+    """
+    components_path = pathlib.Path("components")
+    all_components = [
+        item
+        for item in os.listdir(components_path)
+        if os.path.isdir(components_path / item) and not item.startswith("_")
+    ]
+
+    if not components:
+        # No components were specified at the command line
+
+        if tag_hash == "local":
+            # Build everything if doing a local build
+            components = all_components
+        else:
+            # Only build components that will be published
+            prefix = f"{lamp['name']}-"
+            components = {
+                image[image.startswith(prefix) and len(prefix) :]
+                for image in lamp.get("docker_images", [])
+            }
+
+    # Check that all specified components actually exist in the project
+    for component in components:
+        if component not in all_components:
+            raise ValueError(f"Component '{component}' does not exist")
+
+    if not components:
+        logger.info(
+            "No components found for this project. Create a component directory to get started."
+        )
+        return
+
+    # Check for cycles in the component dependency graph
+    component_graph = validate_component_dependencies(components=components)
+
+    # Let's build in topological order
+    components = [
```

Comment:
this is one case where I *might* use a new variable name, just because it's a little visually confusing to be building this list while filtering based on membership in the same list.